src/gitingest/query_parser.py

```diff
@@ -168,18 +168,47 @@ async def _parse_repo_source(source: str) -> dict[str, Any]:
     parsed["type"] = possible_type
 
     # Commit or branch
-    commit_or_branch = remaining_parts.pop(0)
+    commit_or_branch = remaining_parts[0]
     if _is_valid_git_commit_hash(commit_or_branch):
         parsed["commit"] = commit_or_branch
-    else:
-        parsed["branch"] = commit_or_branch
+        parsed["subpath"] += "/".join(remaining_parts[1:])
 
-    # Subpath if anything left
-    if remaining_parts:
+    else:
+        parsed["branch"] = await _configure_branch_and_subpath(remaining_parts, url)
         parsed["subpath"] += "/".join(remaining_parts)
-
     return parsed
 
+async def _configure_branch_and_subpath(remaining_parts: list[str],url: str) -> str | None:
+    """
+    Find the branch name from the remaining parts of the URL path.
+    Parameters
+    ----------
+    remaining_parts : list[str]
+        List of path parts extracted from the URL.
+    url : str
+        The repository URL to determine branches.
+
+    Returns
+    -------
+    str (branch name) or None
+
+    """
+    try:
+        # Fetch the list of branches from the remote repository
+        branches: list[str] = await fetch_remote_branch_list(url)
+    except Exception as e:
+        print(f"Warning: Failed to fetch branch list: {str(e)}")
```

Comment:
```suggestion
        warnings.warn(f"Failed to fetch branches from {url}: {e}")
```
and `import warnings`